src/ln/channelmanager.rs

```diff
@@ -5362,6 +5365,414 @@ mod tests {
+		create_announced_chan_between_nodes(&nodes, 0, 1);
+
+		let route = nodes[0].router.get_route(&nodes[1].node.get_our_node_id(), None, &Vec::new(), 1000000, TEST_FINAL_CLTV).unwrap();
+		let (payment_preimage_1, payment_hash_1) = get_payment_preimage_hash!(nodes[0]);
+
+		*nodes[0].chan_monitor.update_ret.lock().unwrap() = Err(ChannelMonitorUpdateErr::TemporaryFailure);
+		if let Err(APIError::MonitorUpdateFailed) = nodes[0].node.send_payment(route.clone(), payment_hash_1) {} else { panic!(); }
+		check_added_monitors!(nodes[0], 1);
+
+		let events_1 = nodes[0].node.get_and_clear_pending_events();
+		assert!(events_1.is_empty());
+		assert_eq!(nodes[0].node.list_channels().len(), 1);
+
+		*nodes[0].chan_monitor.update_ret.lock().unwrap() = Ok(());
+		nodes[0].node.test_restore_channel_monitor();
+		check_added_monitors!(nodes[0], 1);
+
+		let mut events_2 = nodes[0].node.get_and_clear_pending_events();
+		assert_eq!(events_2.len(), 1);
+		let payment_event = SendEvent::from_event(events_2.pop().unwrap());
+		assert_eq!(payment_event.node_id, nodes[1].node.get_our_node_id());
+		nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &payment_event.msgs[0]).unwrap();
+		commitment_signed_dance!(nodes[1], nodes[0], payment_event.commitment_msg, false);
+
+		expect_pending_htlcs_forwardable!(nodes[1]);
+
+		let events_3 = nodes[1].node.get_and_clear_pending_events();
+		assert_eq!(events_3.len(), 1);
+		match events_3[0] {
+			Event::PaymentReceived { ref payment_hash, amt } => {
+				assert_eq!(payment_hash_1, *payment_hash);
+				assert_eq!(amt, 1000000);
+			},
+			_ => panic!("Unexpected event"),
+		}
+
+		claim_payment(&nodes[0], &[&nodes[1]], payment_preimage_1);
+
+		// Now set it to failed again...
+		let (_, payment_hash_2) = get_payment_preimage_hash!(nodes[0]);
+		*nodes[0].chan_monitor.update_ret.lock().unwrap() = Err(ChannelMonitorUpdateErr::TemporaryFailure);
+		if let Err(APIError::MonitorUpdateFailed) = nodes[0].node.send_payment(route, payment_hash_2) {} else { panic!(); }
+		check_added_monitors!(nodes[0], 1);
+
+		let events_4 = nodes[0].node.get_and_clear_pending_events();
+		assert!(events_4.is_empty());
+		assert_eq!(nodes[0].node.list_channels().len(), 1);
+
+		// ...and make sure we can force-close a TemporaryFailure channel with a PermanentFailure
+		*nodes[0].chan_monitor.update_ret.lock().unwrap() = Err(ChannelMonitorUpdateErr::PermanentFailure);
+		nodes[0].node.test_restore_channel_monitor();
+		check_added_monitors!(nodes[0], 1);
+
+		let events_5 = nodes[0].node.get_and_clear_pending_events();
+		assert_eq!(events_5.len(), 1);
+		match events_5[0] {
+			Event::BroadcastChannelUpdate { .. } => {},
+			_ => panic!("Unexpected event"),
+		}
+
+		// TODO: Once we hit the chain with the failure transaction we should check that we get a
+		// PaymentFailed event
+
+		assert_eq!(nodes[0].node.list_channels().len(), 0);
+	}
+
+	fn do_test_monitor_temporary_update_fail(disconnect_count: usize) {
```

Comment:
Well... maybe precise the meaning of disconnect_count and disconnect flags, would be easier to review it's a pretty lengthy test here :)

Reply:
Expanded the high-level comment at the top a bunch.